File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/fileInfos/MachO/FatArchInfo.kt
package com.jetbrains.signatureverifier.serialization.fileInfos

import com.jetbrains.signatureverifier.serialization.toByteArray
import kotlinx.serialization.Serializable

@Serializable
sealed class FatArchInfo {
  abstract val cpuType: UInt
  abstract val cpuSubType: UInt

  abstract fun toByteArray(isBe: Boolean): ByteArray
}

@Serializable
data class FatArchInfo64(
  override val cpuType: UInt,
  override val cpuSubType: UInt,
  val fileOffset: ULong,
  val size: ULong,
  val align: ULong,
) : FatArchInfo() {
  override fun toByteArray(isBe: Boolean) =
    cpuType.toByteArray() +
      cpuSubType.toByteArray() +
      fileOffset.toLong().toByteArray(isBe) +
      size.toLong().toByteArray(isBe) +
      align.toLong().toByteArray(isBe)
}

@Serializable
data class FatArchInfo32(
  override val cpuType: UInt,
  override val cpuSubType: UInt,
  val fileOffset: UInt,
  val size: UInt,
  val align: UInt,
) : FatArchInfo() {
  override fun toByteArray(isBe: Boolean) =
    cpuType.toByteArray() +
      cpuSubType.toByteArray() +
      fileOffset.toByteArray(isBe) +
      size.toByteArray(isBe) +
      align.toByteArray(isBe)
}

File: net/JetBrains.FormatRipper/src/MachO/Impl/FatArchInfo.cs
using System;
using JetBrains.FormatRipper.Impl;

namespace JetBrains.FormatRipper.MachO.Impl;

public abstract class FatArchInfo
{
  public uint CpuType;
  public uint CpuSubType;

  public abstract byte[] ToByteArray(bool isBe);
}

public class FatArchInfo64 : FatArchInfo
{
  public ulong FileOffset;
  public ulong Size;
  public ulong Align;

  public FatArchInfo64(uint cpuType, uint cpuSubType, ulong fileOffset, ulong size, ulong align)
  {
    CpuType = cpuType;
    CpuSubType = cpuSubType;
    FileOffset = fileOffset;
    Size = size;
    Align = align;
  }

  public override byte[] ToByteArray(bool isBe) => MemoryUtil.ArrayMerge(
    MemoryUtil.ToByteArray(CpuType, isBe),
    MemoryUtil.ToByteArray(CpuSubType, isBe),
    MemoryUtil.ToByteArray((long)FileOffset, isBe),
    MemoryUtil.ToByteArray((long)Size, isBe),
    MemoryUtil.ToByteArray((long)Align, isBe)
  );
}

public class FatArchInfo32 : FatArchInfo
{
  public uint FileOffset;
  public uint Size;
  public uint Align;

  public FatArchInfo32(uint cpuType, uint cpuSubType, uint fileOffset, uint size, uint align)
  {
    CpuType = cpuType;
    CpuSubType = cpuSubType;
    FileOffset = fileOffset;
    Size = size;
    Align = align;
  }

  public override byte[] ToByteArray(bool isBe) => MemoryUtil.ArrayMerge(
    MemoryUtil.ToByteArray(CpuType, isBe),
    MemoryUtil.ToByteArray(CpuSubType, isBe),
    MemoryUtil.ToByteArray(FileOffset, isBe),
    MemoryUtil.ToByteArray(Size, isBe),
    MemoryUtil.ToByteArray(Align, isBe)
  );
}